cmd/query.go

```diff
@@ -0,0 +1,122 @@
+func newQueryCmd() *cobra.Command {
+	var stack string
+	var configArray []string
+
+	// up implementation used when the source of the Pulumi program is in the current working directory.
+	upWorkingDirectory := func(opts backend.UpdateOptions) result.Result {
+		s, err := requireStack(stack, true, opts.Display, true /*setCurrent*/)
+		if err != nil {
+			return result.FromError(err)
+		}
+
+		// Save any config values passed via flags.
+		if len(configArray) > 0 {
+			commandLineConfig, err := parseConfig(configArray)
+			if err != nil {
+				return result.FromError(err)
+			}
+
+			if err = saveConfig(s, commandLineConfig); err != nil {
+				return result.FromError(errors.Wrap(err, "saving config"))
+			}
+		}
+
+		proj, root, err := readProject()
+		if err != nil {
+			return result.FromError(err)
+		}
+
+		m, err := getUpdateMetadata("" /*msg*/, root)
+		if err != nil {
+			return result.FromError(errors.Wrap(err, "gathering environment metadata"))
+		}
+
+		opts.Engine = engine.UpdateOptions{}
+
+		res := s.Query(commandContext(), backend.UpdateOperation{
+			Proj:   proj,
+			Root:   root,
+			M:      m,
+			Opts:   opts,
+			Scopes: cancellationScopes,
+		})
+		switch {
+		case res != nil && res.Error() == context.Canceled:
+			return result.FromError(errors.New("update cancelled"))
+		case res != nil:
+			return PrintEngineResult(res)
+		default:
+			return nil
+		}
+	}
+
+	// up implementation used when the source of the Pulumi program is a template name or a URL to a template.
+
+	var cmd = &cobra.Command{
+		Use:   "query [template|url]",
+		Short: "Run query program against cloud resources",
+		Long: "Run query program against cloud resources.\n" +
+			"\n" +
+			"This command loads a Pulumi query project and executes it. Pulumi provides various useful constructs\n" +
+			"such as access to stack outputs. All resource operations (create, update, delete) are disallowed in\n" +
+			"query mode." +
+			"\n" +
+			"The program to run is loaded from the project in the current directory by default. Use the `-C` or\n" +
+			"`--cwd` flag to use a different directory.",
+		Args: cmdutil.MaximumNArgs(1),
+		Run: cmdutil.RunResultFunc(func(cmd *cobra.Command, args []string) result.Result {
+			interactive := cmdutil.Interactive()
+
+			opts := backend.UpdateOptions{}
+			opts.Display = display.Options{
+				Color:         cmdutil.GetGlobalColorization(),
+				IsInteractive: interactive,
+				Type:   display.DisplayQuery,
+			}
+
+			return upWorkingDirectory(opts)
+		}),
+	}
+
+	cmd.PersistentFlags().StringVarP(
+		&stack, "stack", "s", "",
+		"The name of the stack to operate on. Defaults to the current stack")
+	cmd.PersistentFlags().StringVar(
+		&stackConfigFile, "config-file", "",
+		"Use the configuration values in the specified file rather than detecting the file name")
+	cmd.PersistentFlags().StringArrayVarP(
+		&configArray, "config", "c", []string{},
+		"Config to use during the update")
```

Comment:
This flag was added to provide a way of updating config for Pulumi apps used with `pulumi up <template>` (as the source for such apps would never be available in a place on disk where you could run `pulumi config set`). We can probably just leave it off of `pulumi query` (`pulumi preview` doesn't have it).